File: src/class035/C7_AllO1.py
class AllOne:
    class Bucket:
        def __init__(self, key, cnt):
            self.set = {key} if key else set()
            self.cnt = cnt
            self.last = None
            self.next = None

    def __init__(self):
        # 创建头尾两个哨兵桶
        self.head = self.Bucket("", 0)
        self.tail = self.Bucket("", float("inf"))
        self.head.next = self.tail
        self.tail.last = self.head
        # 创建哈希表存储键到桶的映射
        self.map = {}

    def insert(self, cur, pos):
        # 在当前桶cur后插入桶pos
        pos.next = cur.next
        pos.next.last = pos
        cur.next = pos
        pos.last = cur

    def remove(self, cur):
        # 移除当前桶cur
        cur.last.next = cur.next
        cur.next.last = cur.last

    def inc(self, key):
        if key not in self.map:
            # 如果key不存在于哈希表中
            if self.head.next.cnt == 1:
                self.map[key] = self.head.next
                self.head.next.set.add(key)
            else:
                new_bucket = self.Bucket(key, 1)
                self.map[key] = new_bucket
                self.insert(self.head, new_bucket)
        else:
            # 如果key已经存在于哈希表中
            bucket = self.map[key]
            if bucket.next.cnt == bucket.cnt + 1:
                self.map[key] = bucket.next
                bucket.next.set.add(key)
            else:
                new_bucket = self.Bucket(key, bucket.cnt + 1)
                self.map[key] = new_bucket
                self.insert(bucket, new_bucket)
            bucket.set.remove(key)
            if not bucket.set:
                self.remove(bucket)

    def dec(self, key):
        if key not in self.map:
            return

        bucket = self.map[key]
        if bucket.cnt == 1:
            del self.map[key]
        else:
            if bucket.last.cnt == bucket.cnt - 1:
                self.map[key] = bucket.last
                bucket.last.set.add(key)
            else:
                new_bucket = self.Bucket(key, bucket.cnt - 1)
                self.map[key] = new_bucket
                self.insert(bucket.last, new_bucket)
        bucket.set.remove(key)
        if not bucket.set:
            self.remove(bucket)

    def getMaxKey(self):
        # 返回计数最高桶中的任意一个键
        if self.tail.last == self.head:
            return ""
        return next(iter(self.tail.last.set))

    def getMinKey(self):
        # 返回计数最低桶中的任意一个键
        if self.head.next == self.tail:
            return ""
        return next(iter(self.head.next.set))

File: src/class035/C7_AllO1.cpp
#include <unordered_map>
#include <unordered_set>
#include <string>
#include <iterator>
using namespace std;

// 全O(1)的数据结构
class AllOne {
    struct Bucket {
        unordered_set<string> set;
        int cnt;
        Bucket* last;
        Bucket* next;

        Bucket(const string& s, int c) : cnt(c), last(nullptr), next(nullptr) {
            set.insert(s);
        }
    };

    // 插入新桶
    void insert(Bucket* cur, Bucket* pos) {
        pos->next = cur->next;
        cur->next->last = pos;
        cur->next = pos;
        pos->last = cur;
    }

    // 删除桶
    void remove(Bucket* cur) {
        cur->last->next = cur->next;
        cur->next->last = cur->last;
        delete cur;
    }

    Bucket* head;
    Bucket* tail;
    unordered_map<string, Bucket*> map;

public:
    AllOne() {
        head = new Bucket("", 0);
        tail = new Bucket("", INT_MAX);
        head->next = tail;
        tail->last = head;
    }

    ~AllOne() {
        Bucket* cur = head;
        while (cur) {
            Bucket* next = cur->next;
            delete cur;
            cur = next;
        }
    }

    void inc(const string& key) {
        if (!map.count(key)) {
            if (head->next->cnt == 1) {
                map[key] = head->next;
                head->next->set.insert(key);
            } else {
                Bucket* newBucket = new Bucket(key, 1);
                map[key] = newBucket;
                insert(head, newBucket);
            }
        } else {
            Bucket* bucket = map[key];
            if (bucket->next->cnt == bucket->cnt + 1) {
                map[key] = bucket->next;
                bucket->next->set.insert(key);
            } else {
                Bucket* newBucket = new Bucket(key, bucket->cnt + 1);
                map[key] = newBucket;
                insert(bucket, newBucket);
            }
            bucket->set.erase(key);
            if (bucket->set.empty()) {
                remove(bucket);
            }
        }
    }

    void dec(const string& key) {
        if (!map.count(key)) return;

        Bucket* bucket = map[key];
        if (bucket->cnt == 1) {
            map.erase(key);
        } else {
            if (bucket->last->cnt == bucket->cnt - 1) {
                map[key] = bucket->last;
                bucket->last->set.insert(key);
            } else {
                Bucket* newBucket = new Bucket(key, bucket->cnt - 1);
                map[key] = newBucket;
                insert(bucket->last, newBucket);
            }
        }
        bucket->set.erase(key);
        if (bucket->set.empty()) {
            remove(bucket);
        }
    }

    string getMaxKey() const {
        if (tail->last == head) return "";
        return *tail->last->set.begin();
    }

    string getMinKey() const {
        if (head->next == tail) return "";
        return *head->next->set.begin();
    }
};
